File: 20.py
# LeetCode 20: Valid Parentases
# Reference: https://www.youtube.com/watch?v=WTzjTskDFMg

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        closeToOpen = {")": "(", "]": "[", "}": "{"}

        for c in s:
            if c in closeToOpen: # c is a closing parentasis
                if stack and stack[-1] == closeToOpen[c]:
                    stack.pop()
                else:
                    return False
                
            else: # c is an opening parentasis
                stack.append(c)

        return True if not stack else False

File: 20.java
// LeetCode 20 Valid Parentases
// Referece: https://www.youtube.com/watch?v=TaWs8tIrnoA

import java.util.Stack;

class Solution {
    public boolean isValid(String s) {

        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()){
            if (c == '('){

                stack.push(')');

            }else if (c == '['){

                stack.push(']');

            } else if (c == '{'){

                stack.push('}');

            } else if (stack.isEmpty() || stack.pop() != c){

                return false;
            }

            }
            return stack.isEmpty();
        }
        
    }
}